sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/CallbackRegistration.java

```diff
@@ -30,50 +33,56 @@
   private final InstrumentDescriptor instrumentDescriptor;
   private final Consumer<T> callback;
   private final T measurement;
-  private final boolean noStoragesRegistered;
+  private final List<AsynchronousMetricStorage<?, ?>> storages;
+  @Nullable private volatile RegisteredReader activeReader;
 
   private CallbackRegistration(
       InstrumentDescriptor instrumentDescriptor,
       Consumer<T> callback,
-      T measurement,
-      List<AsynchronousMetricStorage<?, ?>> storages) {
+      List<AsynchronousMetricStorage<?, ?>> storages,
+      Function<CallbackRegistration<T>, T> measurementProvider) {
     this.instrumentDescriptor = instrumentDescriptor;
     this.callback = callback;
-    this.measurement = measurement;
-    this.noStoragesRegistered = storages.size() == 0;
+    this.measurement = measurementProvider.apply(this);
+    this.storages = storages;
   }
 
   /** Create a {@link CallbackRegistration} for a {@code double} asynchronous instrument. */
   public static CallbackRegistration<ObservableDoubleMeasurement> createDouble(
       InstrumentDescriptor instrumentDescriptor,
       Consumer<ObservableDoubleMeasurement> callback,
       List<AsynchronousMetricStorage<?, ?>> asyncMetricStorages) {
-    ObservableDoubleMeasurement measurement =
-        new ObservableDoubleMeasurementImpl(asyncMetricStorages);
     return new CallbackRegistration<>(
-        instrumentDescriptor, callback, measurement, asyncMetricStorages);
+        instrumentDescriptor,
+        callback,
+        asyncMetricStorages,
+        callbackRegistration -> callbackRegistration.new ObservableDoubleMeasurementImpl());
   }
 
   /** Create a {@link CallbackRegistration} for a {@code long} asynchronous instrument. */
   public static CallbackRegistration<ObservableLongMeasurement> createLong(
       InstrumentDescriptor instrumentDescriptor,
       Consumer<ObservableLongMeasurement> callback,
       List<AsynchronousMetricStorage<?, ?>> asyncMetricStorages) {
-    ObservableLongMeasurement measurement = new ObservableLongMeasurementImpl(asyncMetricStorages);
     return new CallbackRegistration<>(
-        instrumentDescriptor, callback, measurement, asyncMetricStorages);
+        instrumentDescriptor,
+        callback,
+        asyncMetricStorages,
+        callbackRegistration -> callbackRegistration.new ObservableLongMeasurementImpl());
   }
 
   public InstrumentDescriptor getInstrumentDescriptor() {
     return instrumentDescriptor;
   }
 
-  void invokeCallback() {
+  void invokeCallback(RegisteredReader reader) {
```

Comment:
Is this method synchronized in some fashion?

Reply:
At the moment all collections across all readers are synchronized such that only one happens at a time. This is controlled in `MeterSharedState`, which obtains `collectLock` during collection.